IdentityCore/src/cache/mac/MSIDMacCredentialStorageItem.m

```diff
@@ -0,0 +1,220 @@
+// to use, copy, modify, merge, publish, distribute, sublicense, and / or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions :
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+
+#import "MSIDMacCredentialStorageItem.h"
+
+static NSString *keyDelimiter = @"-";
+
+@interface MSIDMacCredentialStorageItem ()
+
+@property (nonatomic) NSMutableDictionary *cacheObjects;
+@property (nonatomic) dispatch_queue_t queue;
+
+@end
+
+@implementation MSIDMacCredentialStorageItem
+
+- (instancetype)init
+{
+    if(self = [super init])
+    {
+        self.cacheObjects = [NSMutableDictionary dictionary];
+        NSString *queueName = [NSString stringWithFormat:@"com.microsoft.universalstorage-%@", [NSUUID UUID].UUIDString];
+        self.queue = dispatch_queue_create([queueName cStringUsingEncoding:NSASCIIStringEncoding], DISPATCH_QUEUE_CONCURRENT);
+    }
+    
+    return self;
+}
+
+- (void)storeCredential:(MSIDCredentialCacheItem *)credential forKey:(NSString *)key
+{
+    dispatch_barrier_async(self.queue, ^{
+        [self.cacheObjects setObject:credential forKey:key];
+    });
+}
+
+- (void)mergeStorageItem:(MSIDMacCredentialStorageItem *)storageItem
+{
+    NSArray *keys = [storageItem allKeys];
+    for (NSString *key in keys)
+    {
+        if (![self storedCredentialForKey:key])
+        {
+            MSIDCredentialCacheItem *credential = [storageItem.cacheObjects objectForKey:key];
+            if (credential)
+            {
+                [self storeCredential:credential forKey:key];
+            }
+        }
+    }
+}
+
+- (MSIDCredentialCacheItem *)storedCredentialForKey:(NSString *)key
+{
+    __block MSIDCredentialCacheItem *credential = nil;
+    
+    dispatch_sync(self.queue, ^{
+        credential = [self.cacheObjects objectForKey:key];
+    });
+    
+    return credential;
+}
+
+- (void)removeStoredCredentialForKey:(NSString *)key
+{
+    dispatch_barrier_async(self.queue, ^{
+        [self.cacheObjects removeObjectForKey:key];
+    });
+}
+
+- (NSArray<MSIDCredentialCacheItem *> *)storedCredentialsForKey:(MSIDDefaultCredentialCacheKey *)key
+{
+    if (key.account && key.service)
+    {
+        NSString *credentialKey = [NSString stringWithFormat:@"%@%@%@", key.account, keyDelimiter, key.service];
```

Comment:
The logic of constructing a credentialKey has been duplicated in both `MSIDMacCredentialStorageItem.m` and `MSIDMacKeychainTokenCache.m`. 
Is it better to have it only here and accept `MSIDDefaultCredentialCacheKey` as a parameter for all APIs in this class? instead of accepting `NSString` key for some APIs and `MSIDDefaultCredentialCacheKey ` key for some other :)

Reply:
Updated.